libcudacxx/docs/extended_api/synchronization_primitives/barrier/arrive_tx.md

```diff
@@ -0,0 +1,64 @@
+---
+grand_parent: Extended API
+parent: Barriers
+---
+
+# `cuda::device::arrive_tx`
+
+Defined in header `<cuda/barrier>`:
+
+```cuda
+__device__
+cuda::barrier<cuda::thread_scope_block>::arrival_token
+cuda::device::arrive_tx(
+  cuda::barrier<cuda::thread_scope_block>& bar,
+  ptrdiff_t arrive_count_update,
+  ptrdiff_t transaction_count_update);
+```
+
+Arrives at a barrier in shared memory, updating both the arrival count and
+transaction count.
+
+## Notes
+
+If `bar` is not in `__shared__` memory, the behavior is undefined. This function
+can only be used under CUDA Compute Capability 9.0 (Hopper) or higher.
+
+To check if `cuda::device::arive_tx` is available, use the
+`__cccl_lib_local_barrier_arrive_tx` feature flag, as shown in the example code below.
+
+## Return Value
+
+An arrival_token, just like the one returned from `cuda::barrier:arrive`.
+
+## Example
+
+```cuda
+#include <cuda/barrier>
+#include <cuda/std/utility> // cuda::std::move
+
+#if defined(__CUDA_MINIMUM_ARCH__) && 900 <= __CUDA_MINIMUM_ARCH__
+#ifndef  __cccl_lib_local_barrier_arrive_tx
+static_assert(false, "Insufficient libcu++ version: cuda::device::arrive_tx is not yet available.");
+#endif // __cccl_lib_local_barrier_arrive_tx
+#endif // __CUDA_MINIMUM_ARCH__
```

Comment:
We should split that into two separate checks
```suggestion
#if defined(__CUDA_MINIMUM_ARCH__) && 900 <= __CUDA_MINIMUM_ARCH__
static_assert(false, "Insufficient CUDA Compute Capability: cuda::device::arrive_tx is not yet available.");
#endif // __CUDA_MINIMUM_ARCH__
#ifndef  __cccl_lib_local_barrier_arrive_tx
static_assert(false, "Insufficient libcu++ version: cuda::device::arrive_tx is not yet available.");
#endif // __cccl_lib_local_barrier_arrive_tx
```

Reply:
The second check will always fail because the feature flag is only available when `900 <= __CUDA_MINIMUM_ARCH__` holds. I don't think this is ideal, so if this documentation example changes anybody's mind, I would be happy to make the feature flag available on all architectures.

Reply:
There is another issue with the current exposure of the feature flag: it triggers too late.

In the following code example with a version of libcu++ that does not yet have the arrive_tx feature
```
#if defined(__CUDA_MINIMUM_ARCH__) && 900 <= __CUDA_MINIMUM_ARCH__
static_assert(false, "Insufficient CUDA Compute Capability: cuda::device::arrive_tx is not yet available.");
#endif // __CUDA_MINIMUM_ARCH__
#ifndef  __cccl_lib_local_barrier_arrive_tx
static_assert(false, "Insufficient libcu++ version: cuda::device::arrive_tx is not yet available.");
#endif // __cccl_lib_local_barrier_arrive_tx

// ..snip..
auto token = cuda::device::arrive_tx(bar, 1, 0);
```
The compiler will error out with message: 
```
error: namespace "cuda::device" has no member "arrive_tx"
```
It will stop compiling before hitting the static_assert. This defeats the purpose of having the check for the feature flag in the first place.

If we do not gate behind `__CUDA_MINIMUM_ARCH__`, then the feature flag check does work: 
```
#ifndef  __cccl_lib_non_existing_feature_flag
static_assert(false, "Insufficient libcu++ version: non-existent feature is not yet available.");
#endif // ! __cccl_lib_non_existing_feature_flag

// .. snip .. 
auto token = cuda::device::arrive_tx(bar, 1, 0);
```
Has error messages: 
```
error: static assertion failed with "Insufficient libcu++ version: non-existent feature is not yet available."
error: namespace "cuda::device" has no member "arrive_tx"
```
Godbolt: https://godbolt.org/z/1sqdc6rc6

@griwes @jrhemstad

Reply:
> The second check will always fail because the feature flag is only available when `900 <= __CUDA_MINIMUM_ARCH__` holds. I don't think this is ideal, so if this documentation example changes anybody's mind, I would be happy to make the feature flag available on all architectures.

Those check different things:

The first check is whether the GPU supports that feature
The second is whether the cccl version is new enough. Someone could use CTK 12.2 and try to get arrive_tx

Reply:
>  The second is whether the cccl version is new enough. Someone could use CTK 12.2 and try to get arrive_tx

I understand. The issue is that the feature flag is not exposed (even when available) when `__CUDA_MIN_ARCH__` is not defined or too small. Therefore, to check whether the feature is available, the check must be surrounded by `#if defined(__CUDA_MINIMUM_ARCH__) && 900 <= __CUDA_MINIMUM_ARCH__ [ ... ] #endif`. This is:
1. Not very user-friendly
2. Hides the static_assert error behind errors about unknown functions.

So as the PR stands currently, a user has two options:

1. Not surround the static_assert with the CUDA_MIN_ARCH check. This will always fail (even when the feature is available) because CUDA_MIN_ARCH is not available during the host part of the compilation trajectory. 
2. Surround the static_assert and feature flag check with the CUDA_MIN_ARCH check. This will work correctly, but will never emit the error message from the static_assert.

Given these two options, we can do three things:
1. Leave things as is. We add a feature flag, but when the feature is not available we leave the user to figure out why `cuda::device::arrive_tx` is unavailable. This to a large extent defeats the purpose of the feature flag.
2. Make the feature flag available regardless of CUDA_MIN_ARCH, so that user code can check if the feature is available in libcu++ (regardless of architecture).
3. Make the feature flag available when either
  a. `__CUDA_MINIMUM_ARCH__` is available and greater or equal to 900.
  b. `__CUDA_MINIMUM_ARCH__` is unavailable.

I just checked that option 3 works ([godbolt](https://godbolt.org/z/TWE37185K)), so going for that one.